probe-rs/src/core/registers.rs

```diff
@@ -293,151 +351,220 @@ impl<T> RegisterValueResultExt<T> for Result<T, Infallible> {
+                    "No stack pointer found. Please report this as a bug.".to_string(),
+                )
+            })
     }
 
     /// The link register.
-    pub fn return_address(&self) -> &RegisterDescription {
-        self.return_address
+    pub fn return_address(&self) -> Result<&CoreRegister, Error> {
+        self.0
+            .iter()
+            .find(|r| r.role == Some(RegisterRole::ReturnAddress))
+            .ok_or_else(|| {
+                Error::GenericCoreError(
+                    "No return address found. Please report this as a bug.".to_string(),
+                )
+            })
     }
 
     /// Returns the nth argument register.
     ///
     /// # Panics
     ///
     /// Panics if the register at given index does not exist.
-    pub fn argument_register(&self, index: usize) -> &RegisterDescription {
-        &self.argument_registers[index]
+    pub fn argument_register(&self, index: usize) -> &CoreRegister {
+        self.get_argument_register(index).unwrap()
     }
 
     /// Returns the nth argument register if it is exists, `None` otherwise.
-    pub fn get_argument_register(&self, index: usize) -> Option<&RegisterDescription> {
-        self.argument_registers.get(index)
-    }
-
-    /// Returns the nth result register.
-    ///
-    /// # Panics
-    ///
-    /// Panics if the register at given index does not exist.
-    pub fn result_register(&self, index: usize) -> &RegisterDescription {
-        &self.result_registers[index]
-    }
-
-    /// Returns the nth result register if it is exists, `None` otherwise.
-    pub fn get_result_register(&self, index: usize) -> Option<&RegisterDescription> {
-        self.result_registers.get(index)
-    }
-
-    /// Returns the nth platform register.
-    ///
-    /// # Panics
-    ///
-    /// Panics if the register at given index does not exist.
-    pub fn platform_register(&self, index: usize) -> &RegisterDescription {
-        &self.platform_registers[index]
-    }
-
-    /// Returns the nth platform register if it is exists, `None` otherwise.
-    pub fn get_platform_register(&self, index: usize) -> Option<&RegisterDescription> {
-        self.platform_registers.get(index)
+    pub fn get_argument_register(&self, index: usize) -> Result<&CoreRegister, Error> {
+        self.0
+            .iter()
+            .filter(|r| matches!(r.role, Some(RegisterRole::Argument(_))))
+            .nth(index)
+            .ok_or_else(|| {
+                Error::GenericCoreError(format!(
+                    "Argument register {index:?} not found. Please report this as a bug."
+                ))
+            })
     }
 
     /// The main stack pointer.
-    pub fn msp(&self) -> Option<&RegisterDescription> {
-        self.msp
+    pub fn msp(&self) -> Result<&CoreRegister, Error> {
+        self.0
+            .iter()
+            .find(|r| r.role == Some(RegisterRole::MainStackPointer))
+            .ok_or_else(|| {
+                Error::GenericCoreError(
```

Comment:
I would rather have this as different Error veriants and they should not be on the top level Error :)

Reply:
@Yatekii OK, but I might need a bit of help if you don't mind. All the other errors in this module are top level (`Error::Other`) errors. How would I go about throwing a lower level error from this level of the code?

Reply:
We could make an `Error::InvalidRegister` and then `RegisterError::NoMainStackPointer` with proper descriptions :) And do this for all variants :)